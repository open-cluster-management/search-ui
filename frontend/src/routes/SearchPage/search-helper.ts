export function formatSearchbarSuggestions(data: string[], suggestionKind: 'label' | 'filter' | 'value', searchQuery: string) {
    let valuesToRemoveFromSuggestions: string[] = []
    const labelTag = {
        id: 'id-suggestions-label',
        key: 'key-suggestions-label',
        name: 'Filters',
        kind: suggestionKind,
        disabled: true
    }
    if (suggestionKind === 'value') {
        // Get a list of duplicate values to remove from suggestions dropdown
        const searchTokens = searchQuery.split(' ')
        const searchCompleteFilter = searchTokens[searchTokens.length - 1].replace(':', '')
        labelTag.name = `${searchCompleteFilter} values`
        const query = convertStringToQuery(searchQuery)
        query.filters.forEach(filter => {
            if (filter.property === searchCompleteFilter) {
                valuesToRemoveFromSuggestions = filter.values.filter(value => data.indexOf(value) > 0)
            }
        })
    }
    const suggestions = data.filter(suggestion => {
        return valuesToRemoveFromSuggestions.indexOf(suggestion) === -1
    }).map((field) => {
        return {
            id: `id-${field}`,
            key: `key-${field}`,
            name: field,
            kind: suggestionKind,
        }
    })
    suggestions.unshift(labelTag)
    return suggestions
}

export const convertStringToQuery = (searchText: string) => {
    const searchTokens = searchText.split(' ')
    const keywords = searchTokens.filter((token) => token !== '' && token.indexOf(':') < 0)
    const filters = searchTokens
        .filter((token) => token.indexOf(':') >= 0)
        .map((f) => {
            const splitIdx = f.indexOf(':')
            const property = f.substring(0, splitIdx)
            const values = f.substring(splitIdx + 1)
            return { property, values: values.split(',') }
        })
        .filter((f) => ['', '=', '<', '>', '<=', '>=', '!=', '!'].findIndex((op) => op === f.values[0]) === -1)
    return {
        keywords,
        filters,
    }
}

export const getSearchCompleteString = (searchQuery: string) => {
    const queryTags = searchQuery.split(' ')
    if (queryTags[queryTags.length - 1].endsWith(':')) {
        return queryTags[queryTags.length - 1].replace(':', '')
    }
    return ''
}