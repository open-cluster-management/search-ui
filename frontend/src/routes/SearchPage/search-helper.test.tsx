import { formatSearchbarSuggestions, convertStringToQuery, getSearchCompleteString } from './search-helper'

test('Correctly returns formatSearchbarSuggestions', () => {
    const testData = ['kind', 'cluster', 'deployment']
    const result = formatSearchbarSuggestions(testData, 'filter', '')
    expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions', () => {
    const testData = ['name1', 'name2', 'name3']
    const result = formatSearchbarSuggestions(testData, 'value', 'name:name1 name:')
    expect(result).toMatchSnapshot()
})

test('Correctly returns convertStringToQuery', () => {
    const testData = 'namespace:open-cluster-management kind:pod'
    const result = convertStringToQuery(testData)
    expect(result).toMatchSnapshot()
})

test('Correctly returns getSearchCompleteString', () => {
    const testData = 'namespace:open-cluster-management kind:'
    const result = getSearchCompleteString(testData)
    expect(result).toMatchSnapshot()
})
