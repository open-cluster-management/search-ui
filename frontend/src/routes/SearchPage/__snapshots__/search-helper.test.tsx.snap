// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly returns convertStringToQuery 1`] = `
Object {
  "filters": Array [
    Object {
      "property": "namespace",
      "values": Array [
        "open-cluster-management",
      ],
    },
    Object {
      "property": "kind",
      "values": Array [
        "pod",
      ],
    },
  ],
  "keywords": Array [],
}
`;

exports[`Correctly returns formatSearchbarSuggestions 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-suggestions-label",
    "key": "key-suggestions-label",
    "kind": "filter",
    "name": "Filters",
  },
  Object {
    "id": "id-kind",
    "key": "key-kind",
    "kind": "filter",
    "name": "kind",
  },
  Object {
    "id": "id-cluster",
    "key": "key-cluster",
    "kind": "filter",
    "name": "cluster",
  },
  Object {
    "id": "id-deployment",
    "key": "key-deployment",
    "kind": "filter",
    "name": "deployment",
  },
]
`;

exports[`Correctly returns formatSearchbarSuggestions 2`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-suggestions-label",
    "key": "key-suggestions-label",
    "kind": "value",
    "name": "name values",
  },
  Object {
    "id": "id-name1",
    "key": "key-name1",
    "kind": "value",
    "name": "name1",
  },
  Object {
    "id": "id-name2",
    "key": "key-name2",
    "kind": "value",
    "name": "name2",
  },
  Object {
    "id": "id-name3",
    "key": "key-name3",
    "kind": "value",
    "name": "name3",
  },
]
`;

exports[`Correctly returns getSearchCompleteString 1`] = `"kind"`;
